src/components/ActionMenu/components/Actions/Actions.tsx

```diff
@@ -35,32 +56,92 @@ export function Actions({actions = [], groups = []}: Props) {
     () => setActiveMenuGroup(undefined),
     [],
   );
-  const menuActions = [...actions, ...groups];
 
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const handleResize = useCallback(
+    debounce(
+      () => {
+        if (!newDesignLanguage || !actionsLayoutRef.current) return;
+        setAvailableWidth(actionsLayoutRef.current.offsetWidth);
+      },
+      20,
+      {leading: false, trailing: true, maxWait: 40},
+    ),
+    [actionsLayoutRef],
+  );
+
+  useEffect(() => {
+    if (!actionsLayoutRef.current) return;
+    setAvailableWidth(actionsLayoutRef.current.offsetWidth);
+  }, [actionsLayoutRef]);
+
+  useEffect(() => {
+    if (!menuGroupRef.current) return;
+    menuGroupWidthRef.current = menuGroupRef.current?.offsetWidth;
+  }, [menuGroupRef]);
+
+  useEffect(() => {
+    if (!newDesignLanguage || actionWidths.length === 0 || availableWidth === 0)
+      return;
+
+    let currentAvailableWidth = availableWidth;
+    setShowableActions([]);
+    setHiddenActions([]);
+
+    actions.forEach((action, index) => {
+      if (
+        actionWidths[index] + menuGroupWidthRef.current + ACTION_SPACING <=
+        currentAvailableWidth
+      ) {
+        currentAvailableWidth -= actionWidths[index];
+        setShowableActions((showableActions) => [
+          ...(showableActions || []),
+          action,
+        ]);
+      } else {
+        currentAvailableWidth = 0;
+        setHiddenActions((hiddenActions) => [...hiddenActions, action]);
+      }
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [availableWidth]);
+
+  const className = classNames(
+    styles.ActionsLayout,
+    newDesignLanguage && styles.newDesignLanguage,
+  );
+
+  const menuActions = [...actions, ...groups];
   const overriddenActions = sortAndOverrideActionOrder(menuActions);
 
   const actionMarkup = overriddenActions.map((action, index) => {
     if ('title' in action) {
       const {title, actions, ...rest} = action;
 
       return actions.length > 0 ? (
-        <MenuGroup
-          key={`MenuGroup-${index}`}
-          title={title}
-          active={title === activeMenuGroup}
-          actions={actions}
-          {...rest}
-          onOpen={handleMenuGroupToggle}
-          onClose={handleMenuGroupClose}
-        />
+        <div key={`MenuGroup-${index}`} ref={menuGroupRef}>
+          <MenuGroup
+            title={title}
+            active={title === activeMenuGroup}
+            actions={[...hiddenActions, ...actions]}
+            {...rest}
+            onOpen={handleMenuGroupToggle}
+            onClose={handleMenuGroupClose}
+          />
+        </div>
       ) : null;
     }
-
+    if (showableActions && showableActions.length >= 0) return null;
```

Comment:
If there are showable actions, we don't render this below but it's necessary to render them at first to get their widths